python/cudf/cudf/_libxx/join.pyx

```diff
@@ -4,58 +4,100 @@ from cudf._libxx.includes.join cimport *
 from libcpp.vector cimport vector
 from libcpp.pair cimport pair
 
-cpdef join(lhs, rhs, left_on, right_on, how, method):
+from collections import OrderedDict
+
+cpdef join(lhs, rhs, left_on, right_on, how, method, left_index=False, right_index=False):
     """
     Call libcudf++ join for full outer, inner and left joins.
     Returns a list of tuples [(column, valid, name), ...]
     """
-    if how not in ['left', 'inner', 'outer']:
-        msg = "new join api only supports left, inner or outer"
-        raise ValueError(msg)
-
-    missing_key_err = 'column "{}" not found in {}'
-    for name in left_on:
-        if name not in lhs._data.keys():
-            raise KeyError(missing_key_err.format(name, 'left'))
-    for name in right_on:
-        if name not in rhs._data.keys():
-            raise KeyError(missing_key_err.format(name, 'right'))
-
-    left_idx = []
-    right_idx = []
-
-    assert(len(left_on) == len(right_on))
 
     cdef vector[int] left_on_ind
     cdef vector[int] right_on_ind
     cdef vector[pair[int, int]] columns_in_common
 
-    result_col_names = []  # Preserve the order of the column names
+    # convert both left and right to c table objects
+    cdef Table c_lhs = lhs
+    cdef Table c_rhs = rhs
+
+    # create the two views that will be set later to include or not include the index column 
+    cdef table_view lhs_view
+    cdef table_view rhs_view
+
+    # the result columns are all the left columns (including common ones) 
+    # + all the right columns (excluding the common ones)
+    result_col_names = []
 
-    for name, col in lhs._data.items():
+    for name in lhs._data.keys():
```

Comment:
Could this be just `result_col_names = lhs._data.keys() | rhs._data.keys()`?

Reply:
I wish. The issue is that the resulting set's order isn't guaranteed. Here libcudf returns the columns starting with all those on the left, including the common columns, followed by all those on the right, excluding the common columns. We need to pick those columns up with exactly the right labels. Here's a repro:
```
>>> d1 = {'a':1, 'b':2, 'c':3, 'd':4, 'e':5}
>>> d2 = {'f':6, 'g':7, 'h':8}
>>> d1.keys() | d2.keys()
{'a', 'f', 'c', 'd', 'e', 'g', 'h', 'b'}
```

Reply:
Ah, the set doesn't preserve order. Bummer!